### PR TITLE
jit pickling rref

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32959 jit pickling rref**

in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit e08cae3d:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=e08cae3d0894eda05f90c719ddd01af6d5180173).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4407466) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94828830) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94828830))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
do so (now or later) by using -b with the checkout command again. Example: 
 
  git checkout -b <new-branch-name> 
 
HEAD is now at e08cae3d08 jit pickling rref 
+ git reset --hard e08cae3d0894eda05f90c719ddd01af6d5180173 
HEAD is now at e08cae3d08 jit pickling rref 
+ git merge --allow-unrelated-histories --no-edit --no-ff 5ca7bf453df9b5854062dc581ffd398f6b1b2b30 
Auto-merging torch/testing/_internal/distributed/rpc/rpc_test.py 
CONFLICT (content): Merge conflict in torch/testing/_internal/distributed/rpc/rpc_test.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/csrc/jit/pickler.cpp
Comment (pritamdamania87) [inline torch/csrc/jit/pickler.cpp]: I know this is what PyRRef does, but as the comment there mentions this approach is error prone. If we jit serialize a model containing an RRef we would end up creating a child fork incorrectly here.

Can we ensure we create a child fork only when we're sending RPCs?

@mrshenli Do we have gh issue tracking that we need to fix this for PyRRef?

File: torch/csrc/jit/unpickler.cpp
Comment (pritamdamania87) [inline torch/csrc/jit/unpickler.cpp]: nit: Should we add an assert for the number of args we expect?

File: torch/csrc/jit/unpickler.cpp
Comment (pritamdamania87) [inline torch/csrc/jit/unpickler.cpp]: Similar to the pickle method, this is error prone when we jit unpickle a module that contains an RRef. In general pickling/unpickling should not have any RRef fork logic, only the RPC Framework should have this logic.

File: torch/csrc/distributed/rpc/rref_context.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/rref_context.cpp]: I think it might be fine to patch this PR for the performance effort to unblock them. But I don't think we should land a PR with a change like this where we remove an important assert.

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/rpc/rpc_test.py]: @mrshenli How do other RRef tests guarantee that RRefs are garbage collected before shutdown() is called?

Comment (mrshenli) [reply torch/testing/_internal/distributed/rpc/rpc_test.py]: Other tests rely on Python to deref local RRef objects when exiting the test function context, and then uses agent.join() to wait for all messages settle.

File: torch/csrc/jit/pickler.cpp
Comment (mrshenli) [reply torch/csrc/jit/pickler.cpp]: Just created one #32966

**Comment by mrshenli** [comment]
@zhaojuanmao could you please confirm that the leak you encountered in this PR did not occur in #32674?

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/rpc_test.py]: IIUC, this will return a PyRRef that contains an intrusive_ptr of RRef. So, when you pass this as an arg to `rpc.rpc_sync`, how does `createStackForSchema` work? Does it know it should skip the pickler of PyRRef?

Comment (mrshenli) [reply torch/testing/_internal/distributed/rpc/rpc_test.py]: How does pickler serialize `PyObjectHolder`?

File: torch/csrc/distributed/rpc/rref_context.cpp
Comment (zhaojuanmao) [reply torch/csrc/distributed/rpc/rref_context.cpp]: agree, as commented here, this is mainly to unblock test, not for landing diff

File: torch/csrc/jit/pickler.cpp
Comment (zhaojuanmao) [reply torch/csrc/jit/pickler.cpp]: So far the build and test do not complain, but if there is another approach to avoid dependency on rref_context, that is even better, will follow up with you offline

File: torch/csrc/jit/pickler.cpp
Comment (zhaojuanmao) [reply torch/csrc/jit/pickler.cpp]: agree this is an issue we need to tackle. but I think it is better tackled with python pickling together, as in general JIT pickling serialization format should be same as Python pickling serialization format

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (zhaojuanmao) [reply torch/testing/_internal/distributed/rpc/rpc_test.py]: @mrshenli I checked the code, it worked like this:
when we define torchscript function run_ref_script_module(), its input type is annotated, so when we call createStackForSchema(), JIT knows the input type is RRef type, then when it reaches toIvalue of pybind_utils.h, it will treat the argument as RRef type and call obj.cast<torch::distributed::rpc::PyRRef>().toIValue() => this will return c10::static_intrusive_pointer_cast<c10::RRefInterface>(rref_), which is the IValue we want

**Comment by zhaojuanmao** [comment]
> @zhaojuanmao could you please confirm that the leak you encountered in this PR did not occur in #32674?

Yes, the leak did not happen in #32674

File: torch/csrc/jit/pickler.cpp
Comment (driazati) [inline torch/csrc/jit/pickler.cpp]: I'm not very familiar with the semantics of RRef are but the changes to the pickler here look fine. Can you elaborate on what it means to serialize an RRef? Is it possible to load a model much later and have a dead reference or something?

Comment (zhaojuanmao) [reply torch/csrc/jit/pickler.cpp]: thanks for reviewing @driazati. We are implementing how to serialize RRef during RPC call only right now, we have not implemented how to save a model with RRef. 

When we make RPC call that takes RRef as args, we need to pickle RRef by passing a few IDs and type of the RRef over the wire, and then recreate the RRef from the received IDs and type during unpickling.

I will add more description in the summary

File: torch/csrc/jit/pickler.cpp
Comment (pritamdamania87) [reply torch/csrc/jit/pickler.cpp]: This is a pretty serious bug during pickling and unpickling. Can we fix this the right way in this PR?

**Comment by suo** [comment]
Checking in here: what is the status of this? Is it blocked on waiting for review? If so, who is the best reviewer? Is it me? 😛

**Comment by zhaojuanmao** [comment]
@suo thanks for checking. Yes, this diff is blocked by reviewing from both JIT team and our team. 

That would be great if you can help review whether you are OK with JIT unpickler constructor change, class_resolver does not work for our case, I added a type_resolver to JIT unpickler.  

@driazati commented he is fine with where I put rref pickling logic to.

**Comment by zhaojuanmao** [comment]
updated summary in git commit message (did not show in the PR summary somehow) and updated some comments

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (mrshenli) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: any reason for creating a new lambda every time calling this? This will be on the critical path as we need this for every message. Is it possible to have a static var here, and share it through `shared_ptr` or reference?

File: torch/csrc/distributed/rpc/py_rref.cpp
Comment (mrshenli) [inline torch/csrc/distributed/rpc/py_rref.cpp]: Thanks for fixing this!

File: torch/csrc/distributed/rpc/rpc_agent.h
Comment (mrshenli) [inline torch/csrc/distributed/rpc/rpc_agent.h]: instroduce => introduce

File: torch/csrc/jit/pickler.cpp
Comment (mrshenli) [reply torch/csrc/jit/pickler.cpp]: One option would be having a thread_local flag/context. We use this context/flag to wrap the code below (and for remote as well, basically every rpc related entry point). Then the pickler can check the thread_local context/flag. If not set, throw an error.

https://github.com/pytorch/pytorch/blob/d613bd0522d51602ccf8cb29d00eb26cf3af923d/torch/distributed/rpc/api.py#L457-L465

**Review by mrshenli — COMMENTED** [review]
For the leak, I set `api._ignore_rref_leak = False`, and run `test_remote_script_module` 60 times, but cannot reproduce the leak. Do you still see the leak in your dev env?

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: Why is this part of the RpcAgent interface? It seems like this can just be a static utility function in rpc/utils.cpp?

File: torch/csrc/jit/pickle.h
Comment (pritamdamania87) [inline torch/csrc/jit/pickle.h]: Can we add some comments to the pickle and unpickle APIs in terms of what the type_resolve is?

File: torch/csrc/jit/pickler.cpp
Comment (pritamdamania87) [inline torch/csrc/jit/pickler.cpp]: Use TORCH_CHECK here or C10_THROW_ERROR. TORCH_INTERNAL_ASSERT will show up as a bug to the user and ask them to file a gh issue.

Also, what was the reason we needed to add this #ifdef? If USE_DISTRIBUTED is false, the user can't even create RRefs if I understand correctly, so how will they even pickle/unpickle one?

File: torch/csrc/jit/pickler.cpp
Comment (pritamdamania87) [inline torch/csrc/jit/pickler.cpp]: Why do we need this ifdef?

File: torch/csrc/jit/pickler.cpp
Comment (pritamdamania87) [reply torch/csrc/jit/pickler.cpp]: Yanli and I spoke offline and thought its better to deal with this in a follow up PR. Although, lets fix this issue soon as a follow up PR for both jit serialization and PyRRef serialization.

The thread local idea sounds good since it can be reused across python pickle and JIT pickle.

File: torch/csrc/jit/pickler.h
Comment (pritamdamania87) [inline torch/csrc/jit/pickler.h]: Why do we need ifdef here? This is just a function declaration.

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/rpc/rpc_test.py]: Nice unit tests!

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]
Requesting changes mostly for the RpcAgent interface changes.

**Comment by suo** [comment]
Okay I am going to resign from reviewing until the other reviews are sorted out, just to avoid too much cross-chatter. Please re-request when you would me to take a look.

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (zhaojuanmao) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: synced up offline, this is because we want to avoid introducing python dependency for files who call the getTypeResolver(), will create two new files: one file will have virtual getTypeResolver() interface that can be called by callers without python dependency, another file will have getTypeResolver() implementation that could have python dependency

File: torch/csrc/jit/pickler.cpp
Comment (zhaojuanmao) [reply torch/csrc/jit/pickler.cpp]: synced up offline, USE_DISTRIBUTED is needed here as for some use cases, users do not need to compile rref_context.h if they are not building distributed package

File: torch/csrc/jit/pickler.cpp
Comment (zhaojuanmao) [reply torch/csrc/jit/pickler.cpp]: same as above

File: torch/csrc/jit/pickler.cpp
Comment (zhaojuanmao) [reply torch/csrc/jit/pickler.cpp]: sounds good to me as well

File: torch/csrc/jit/pickler.h
Comment (zhaojuanmao) [reply torch/csrc/jit/pickler.h]: same as above

File: torch/csrc/jit/unpickler.h
Comment (suo) [inline torch/csrc/jit/unpickler.h]: I don't think we should have a separate `ClassResolver` and `TypeResolver`, since they do the same job (take a type name and return the TypePtr). Instead, we should modify `StrongTypePtr` to have the `cu_` field be nullable (and enforce that if the `type_` field is a class, then `cu_` must be non-null). That way, we can merge the two resolvers.

Aside: the name of merged resolver should be `TypeResolver`, since that best describes it.

**Review by suo — COMMENTED** [review]
I have one simplification suggestion. Otherwise, if @driazati signed off on the actual pickling/unpickling code, I think the JIT side is good to go.

File: torch/csrc/jit/unpickler.h
Comment (zhaojuanmao) [reply torch/csrc/jit/unpickler.h]: sounds good!

File: torch/csrc/distributed/rpc/rpc_agent.h
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/rpc_agent.h]: nit: The typeResolver_ and all the get/set methods can be static right?

**Review by pritamdamania87 — APPROVED** [review]
Approving since @suo mentioned the JIT side is good to go.

**Pull request closed without merging**
